packages/superset-ui-core/test/query/processExtraFormData.test.ts

```diff
@@ -0,0 +1,183 @@
+        {
+          datasource: 'table_1',
+          granularity: 'something',
+          viz_type: 'custom',
+        },
+        {
+          filters: [{ col: 'my_col', op: '==', val: 'my value' }],
+        },
+      ),
+    ).toEqual({
+      datasource: 'table_1',
+      filters: [{ col: 'my_col', op: '==', val: 'my value' }],
+      granularity: 'something',
+      viz_type: 'custom',
+    });
+  });
+
+  it('should add allowed values to pre-existent value', () => {
+    expect(
+      appendExtraFormData(
+        {
+          granularity: 'something',
+          viz_type: 'custom',
+          datasource: 'table_1',
+          filters: [{ col: 'my_col', op: '==', val: 'my value' }],
+        },
+        {
+          filters: [{ col: 'my_other_col', op: '!=', val: 'my other value' }],
+        },
+      ),
+    ).toEqual({
+      granularity: 'something',
+      viz_type: 'custom',
+      datasource: 'table_1',
+      filters: [
+        { col: 'my_col', op: '==', val: 'my value' },
+        { col: 'my_other_col', op: '!=', val: 'my other value' },
+      ],
+    });
+  });
+
+  it('should add new freeform where', () => {
+    expect(
+      appendExtraFormData(
+        {
+          datasource: 'table_1',
+          granularity: 'something',
+          viz_type: 'custom',
+        },
+        {
+          extras: {
+            where: '1 = 0',
+          },
+        },
+      ),
+    ).toEqual({
+      datasource: 'table_1',
+      granularity: 'something',
+      viz_type: 'custom',
+      extras: {
+        where: '(1 = 0)',
+      },
+    });
+  });
+
+  it('should add new freeform where to existing where clause', () => {
+    expect(
+      appendExtraFormData(
+        {
+          datasource: 'table_1',
+          granularity: 'something',
+          viz_type: 'custom',
+          extras: {
+            where: 'abc = 1',
+          },
+        },
+        {
+          extras: {
+            where: '1 = 0',
+          },
+        },
+      ),
+    ).toEqual({
+      datasource: 'table_1',
+      granularity: 'something',
+      viz_type: 'custom',
+      extras: {
+        where: '(abc = 1) AND (1 = 0)',
+      },
+    });
+  });
+});
+
+describe('overrideExtraFormData', () => {
+  it('should assign allowed non-existent value', () => {
+    expect(
+      overrideExtraFormData(
+        {
+          granularity: 'something',
+          viz_type: 'custom',
+          datasource: 'table_1',
+        },
+        {
+          time_grain_sqla: 'PT1H',
+        },
+      ),
+    ).toEqual({
+      granularity: 'something',
+      viz_type: 'custom',
+      datasource: 'table_1',
+      time_grain_sqla: 'PT1H',
+    });
+  });
+
+  it('should override allowed pre-existing value', () => {
+    expect(
+      overrideExtraFormData(
+        {
+          granularity: 'something',
+          viz_type: 'custom',
+          datasource: 'table_1',
+          time_grain_sqla: 'PT1H',
+        },
+        {
+          time_grain_sqla: 'PT2H',
```

Comment:
This makes me curious, what happens/explodes if you `appendExtraFormData` with these inputs?

Reply:
In this case nothing happens, as they're not on the allow list. I decided to explicitly list all allowed overrides/appends, as changing e.g. the dataset might have security implications.